File: tgbot/internal/auth_manager.go
// auth_manager.go
package internal

import (
	"sync"
	. "tgbot/pkg"
)

type AuthManager struct {
	players map[int64]*Player
	mu      sync.Mutex
}

func NewAuthManager() *AuthManager {
	return &AuthManager{players: make(map[int64]*Player)}
}

// RegisterUser добавляет нового пользователя с паролем
func (am *AuthManager) RegisterUser(telegramID int64, name, password string) *Player {
	am.mu.Lock()
	defer am.mu.Unlock()

	player := NewPlayer(name, telegramID, password)
	am.players[telegramID] = player
	return player
}

// Authenticate проверяет, зарегистрирован ли пользователь и совпадает ли пароль
func (am *AuthManager) Authenticate(telegramID int64, name, password string) bool {
	am.mu.Lock()
	defer am.mu.Unlock()

	player, exists := am.players[telegramID]
	return exists && player.Password == password
}

// GetPlayer возвращает игрока по его Telegram ID
func (am *AuthManager) GetPlayer(telegramID int64) *Player {
	am.mu.Lock()
	defer am.mu.Unlock()

	return am.players[telegramID]
}

// IsRegistered проверяет, зарегистрирован ли пользователь
func (am *AuthManager) IsRegistered(telegramID int64) bool {
	am.mu.Lock()
	defer am.mu.Unlock()

	_, exists := am.players[telegramID]
	return exists
}

File: tgbot/cppbot/auth_manager.cpp
#include <unordered_map>
#include <string>
#include <memory>
#include "models.h"

class AuthManager {
public:
    bool register_user(int64_t telegram_id, const std::string& name);
    std::shared_ptr<Player> get_player(int64_t telegram_id) const;
    bool authenticate(int64_t telegram_id) const;

private:
    std::unordered_map<int64_t, std::shared_ptr<Player>> players;
};

bool AuthManager::register_user(int64_t telegram_id, const std::string& name) {
    if (players.find(telegram_id) == players.end()) {
        auto player = std::make_shared<Player>(name, telegram_id);
        players[telegram_id] = player;
        return true;
    }
    return false;
}

std::shared_ptr<Player> AuthManager::get_player(int64_t telegram_id) const {
    auto it = players.find(telegram_id);
    return (it != players.end()) ? it->second : nullptr;
}

bool AuthManager::authenticate(int64_t telegram_id) const {
    return players.find(telegram_id) != players.end();
}